### ISSUE TITLE
Suggestion: remove ^c--possibly entirely, or at least from non-GCU front-ends

### ISSUE THREAD

**Issue by smbhax** [angband/angband#5542]
Clunking around on my new keyboard with slightly differently sized keys, I've found myself accidentally hitting ^c instead of ^x a lot--so instead of a confirmation that I hit Save and am quitting appearing, a prompt checking to see if I actually want to delete my character appears. = o

I suppose hitting "y" from there actually to delete the character is a stretch, and I haven't fumbled onto that yet, but it still feels hazardous every time.

I asked on the forum if that was supposed to happen ("?" says ^c is "special - break" but I didn't know what that meant; all I could see was that ^c has a regenerating keymap to Q, which is "kill character and quit"(!)), and Nick [explained  ](http://angband.oook.cz/forum/showthread.php?p=159585#post159585)that

> Back in the old days before Microsoft started using Ctrl-C for copy, it was the key combination to stop a running program. Rather than actually allowing that to happen, Angband catches the key combination and asks the player if they're trying to kill the character. 

Given that it probably doesn't apply to newer systems, then, is it worth preserving this potential kill-the-player function, especially given that it's the key right next to Save on the keyboard? Could ^c be removed from the non-GCU front ends?

Or... Is it even worth having a function to MAYBE assist the user in killing their player via a non-Angband command anyway? Angband has Q for that. What about just dropping the ^c kill prompt entirely? Continue preventing ^c from doing program-stop, and rely on the player using the proper Angband Q command if that's really what they want to do. Why cater to a duplicate and game-incorrect input with a function that is a constant permadeath hazard to all characters on all systems for particularly clumsy and hasty users such as myself?

**Comment by backwardsEric** [comment]
A workaround, until we decide what to do the ctrl-c keymap in pref.prf, would be to install your own keymap for ctrl-c (not possible in the GCU front end) so it does something harmless and save it in a preference file so it will be reloaded when the character is reloaded (i.e. like the recommendation, https://angband.readthedocs.io/en/latest/faq.html#is-there-a-way-to-disable-that-thing-that-pops-up-when-you-hit-the-enter-key , for how to disable the enter menu).

As for preventing ctrl-c from killing the program for the GCU front end, I'd lean to leaving it as it is.  In general, ctrl-c is intended to kill an unresponsive process, and Angband's behavior was probably intended to keep that functionality while trying to prevent it from being abused as a method for save scumming (facing imminent death or some other disagreeable situation, ctrl-c to interrupt, and then reload the character at the point it was last saved).  Though, unless one's using a setgid version, there's plenty of ways to save scum since the player has ready access to the save file.  So, perhaps ctrl-c should only behave the way it currently does in the GCU front end for a setgid version.

**Comment by smbhax** [comment]
Ah, thanks! I made a blank keymap over it (didn't even need the Space mentioned in the manual), saved to user.prf, that sure did the trick. : D

**Comment by backwardsEric** [comment]
So my current thoughts on this are to drop the ctrl-c keymap in pref.prf and change the signal handler, that pressing ctrl-c in the GCU frontend triggers, so the current behavior after ctrl-c has been pressed enough of suiciding the character becomes a quit without save for a non-setgid installation and retains the current behavior for a setgid installation.

**Comment by smbhax** [comment]
I'm not entirely sure what that means, but if it means CTRL+C still effectively kills the character in MSYS2 SDL2, to me that's still a problem--and seems unnecessary. I guess I'm Windows-centric, and I don't know if actual Linux is similar, but in Windows you have Task Manager if you need to kill an unresponsive process. The normal use of CTRL+C in Windows is "Copy" for the system clipboard; I don't think any Windows users are going to be hitting CTRL+C expecting it to kill the program. A kill button right next to the game's "Save & Quit" button is a constant hazard.

**Comment by backwardsEric** [comment]
For the SDL2 front end where Angband has keyboard focus and was waiting for a command, ctrl-c with that change would generate the message for an unrecognized command (since there's no default keymap for it in pref.prf).  Now, at least on Linux or other Unix/Unix-like system if you typed ctrl-c in the terminal that started Angband (and had not run Angband in the background) then the terminal would relay that as a signal to Angband and that (after hitting ctrl-c enough) would trigger Angband to exit.  To me that seems reasonable, since hitting ctrl-c in a terminal is supposed to interrupt what the terminal is currently running.

**Comment by smbhax** [comment]
Make sense.

**Comment by smbhax** [comment]
I forgot that you told me stopping ^c with a blank keymap wasn't possible in the GCU front end, and ended up killing myself several times in trying to set up such a keymap--because the key was getting intercepted and I couldn't figure out how to type it in a keymap and well obviously that just wasn't going to work.

In flailing around with it I did notice a few odd things though:

For one thing, it hijacks the cursor position. So if say you're in the keymap editor, defining a keymap, and you hit ^c four times, the first three times you get nothing, then the fourth time you get this weird, not very clear text coming up up above saying just "Contemplating suicide!"

![Untitled-1 copy](https://user-images.githubusercontent.com/91744319/203026768-02ae6c88-22d1-48b8-86c8-df1baa5f9fbd.jpg)

- It DOESN'T really make it super clear that hitting ^c again (the fifth consecutive time) will kill you; yeah that could be guessed but shouldn't this be really clear?
- It moves the cursor position up to the "Contemplating suicide!" line, so the next output, like from the keymap entry routine if you continue with that, will pop out up there, instead of down where the "Key:" line is.

Second, in trying to make sense of this (still having forgotten your mentioning the keymap override doesn't work in GCU, and finding it IS definitely a hardcoded keymap in pref.prf but for the reason I didn't understand then (it's being intercepted by the terminal or whatever) couldn't be queried in the keymaps menu), I googled and found the manual section on keymaps

https://angband.readthedocs.io/en/latest/customize.html#keymaps

which appears to be quite out of date: 

- capitalized letters are used for indicating what are actually non-capitalized letters, like ^[letter] (and "^c" is written in two spots as "^C" in pref.prf)
- the manual gives the wrong menu location for the keymaps menu: "To set up keymaps, go to the options menu (=) and select “Edit keymaps” (k)."
- the content here is very unclear and not clearly applicable to using the commands available in the current keymaps menu

And yeah I would say the two instances of the "#Hack -- Commit suicide" macro in pref.prf should definitely be dropped; there's no point in duplicating that horrific functionality in other front ends, and anyway it's morbidly deceptive when it's being called "Commit suicide" in the prf but the "?" key listing just vaguely refers to it as "special - break."

But eh hm well for me, a noob who doesn't really understand terminals, I would still stupidly kind of like to not have ^c being any kind of suicide button. Like, it's not really truly operating like a straight interrupt anymore anyway, so why not at least save first before quitting, instead of killing the character?

**Comment by smbhax** [comment]
Looking at ui_signals.c, it looks like ^c is supposed to trigger a noise when pressed the second, third, and fourth times. Mine doesn't make a noise.

Anyway I don't think this is something I need, it's what Windows has Task Manager for. I've commented my ui-signals.c line 121 to disable it.

**Comment by backwardsEric** [comment]
As for the noise, you might try replacing the "case TERM_XTRA_NOISE: ..." line (line 1063 in the current version) with "case TERM_XTRA_NOISE: beep(); return 0;" and then recompiling the GCU front end.  That uses the curses/ncurses call to ring the terminal's bell and may be more portable than what's there.  One way to check if that's working is to enter targeting/look mode and press a key that's not understood by the targeting code (neither a direction nor anything in '-pomt50.g?' + the key for ignoring an object in the current keyset).

**Comment by smbhax** [comment]
(Tried that main-gcu line replacement but no still beeps in Cygwin, neither when hitting wrong keys in targeting mode nor when hitting ^c four times after re-enabling it (couldn't figure out what "-pomt50.g" means = o; I tried keys that were neither move keys nor keys listed in targeting mode's ? text).)

**Comment by smbhax** [comment]
OH. Cygwin "Options - Terminal - Bell - Wave" was set to "none (system sound)"; when I point it to a sound file name in the drop-down list, yes I do get sounds in targeting mode with invalid keys (didn't re-enable ^c to check that), using the stock main-gcu.c.

Not sure if "none (system sound)" is Cygwin's default, or if I turned it off early on because I found beeps annoying. I don't remember turning it off, but I'm not sure.

... It is annoying though so I guess I'll set it back to "none." = o

But anyway yeah Angband's beep works, maybe defeated by stupid users or terminal default settings.
